modules/javafx.base/src/test/java/test/javafx/beans/value/ObservableValueWhenTest.java

```diff
@@ -0,0 +1,357 @@
+                assertEquals(List.of("a -> b", "b -> c", "c -> d"), observedChanges);
+            }
+        }
+
+        @Nested
+        class AndConditionStartsTrue {
+            BooleanProperty condition = new SimpleBooleanProperty(true);
+
+            @Test
+            void shouldCallDownstreamMapFunctionOnlyWhenAbsolutelyNecessary() {
+                StringProperty property = new SimpleStringProperty("a");
+                List<String> observedMappings = new ArrayList<>();
+                List<String> observedChanges = new ArrayList<>();
+
+                property.when(condition)
+                    .map(x -> { observedMappings.add(x); return x; })
+                    .addListener((obs, old, current) -> observedChanges.add(old + " -> " + current));
+
+                assertEquals(List.of("a"), observedMappings);
+                assertEquals(List.of(), observedChanges);
+
+                property.set("b");
+
+                assertEquals(List.of("a", "b"), observedMappings);
+                assertEquals(List.of("a -> b"), observedChanges);
+
+                condition.set(false);
+
+                assertEquals(List.of("a", "b"), observedMappings);
+                assertEquals(List.of("a -> b"), observedChanges);
+
+                property.set("c");
+
+                assertEquals(List.of("a", "b"), observedMappings);
+                assertEquals(List.of("a -> b"), observedChanges);
+
+                condition.set(true);
+
+                assertEquals(List.of("a", "b", "c"), observedMappings);
+                assertEquals(List.of("a -> b", "b -> c"), observedChanges);
+
+                property.set("d");
+
+                assertEquals(List.of("a", "b", "c", "d"), observedMappings);
+                assertEquals(List.of("a -> b", "b -> c", "c -> d"), observedChanges);
+
+                condition.set(false);
+
+                assertEquals(List.of("a", "b", "c", "d"), observedMappings);
+                assertEquals(List.of("a -> b", "b -> c", "c -> d"), observedChanges);
+            }
+        }
+    }
+
+    @Nested
+    class WhenObservedDirectlyForInvalidations {
+
+        @Nested
+        class AndConditionStartsFalse {
+            BooleanProperty condition = new SimpleBooleanProperty(false);
+
+            @Test
+            void shouldOnlyInvalidateWhenAbsolutelyNeeded() {
+                StringProperty property = new SimpleStringProperty("a");
+                AtomicInteger observedInvalidations = new AtomicInteger();
+
+                ObservableValue<String> when = property.when(condition);
+
+                when.addListener(obs -> observedInvalidations.addAndGet(1));
+
+                assertEquals(0, observedInvalidations.get());
+
+                property.set("b");
+
+                assertEquals(0, observedInvalidations.get());
+
+                when.getValue();  // would make no difference, inactive when bindings are always valid
```

Comment:
:-) The confusing part is because I referred to the `when` construct here.  I've changed it to:

      // would make no difference, inactive "when" bindings are always valid